src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -110,54 +110,148 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
-      val nextCursor = sr.nextCursor.getOrElse(Map.empty).get("sync")
+  private def generateTableExpression(
+      instanceType: InstanceType,
+      filters: FilterDefinition): TableExpression =
+    instanceType match {
+      case InstanceType.Edge =>
+        TableExpression(edges = Some(EdgeTableExpression(filter = Some(filters))))
+      case InstanceType.Node =>
+        TableExpression(nodes = Some(NodesTableExpression(filter = Some(filters))))
+    }
+
+  private def fetchData(
+      isBackFill: Boolean,
+      cursors: Option[Map[String, String]],
+      `with`: Map[String, TableExpression],
+      select: Map[String, SelectExpression]): IO[ItemsWithCursor[InstanceDefinition]] = {
+    val response = if (isBackFill) {
+      client.instances.queryRequest(
+        InstanceQueryRequest(
+          `with` = `with`,
+          cursors = cursors,
+          select = select,
+          includeTyping = Some(true)
+        )
+      )
+    } else {
+      client.instances.syncRequest(
+        InstanceSyncRequest(
+          `with` = `with`,
+          cursors = cursors,
+          select = select,
+          includeTyping = Some(true)
+        )
+      )
+    }
+    response.map { qr =>
+      val itemDefinitions = qr.items.getOrElse(Map.empty).getOrElse("sync", Vector.empty)
+      val nextCursor = qr.nextCursor.getOrElse(Map.empty).get("sync")
       ItemsWithCursor(itemDefinitions, nextCursor)
     }
   }
+
+  private def syncOut(
+      syncMode: SyncMode,
+      `with`: Map[String, TableExpression],
+      select: Map[String, SelectExpression],
+      selectedProps: Array[String])(
+      implicit F: Async[IO]): Stream[IO, ProjectedFlexibleDataModelInstance] = {
+    val (isBackFill, cursors, toProjectedCursor) = syncMode match {
+      case StreamMode(syncCursors) =>
+        (false, Some(syncCursors), (nextCursor: Option[String]) => nextCursor)
+      case BackFillMode(futureItemsSyncCursor) =>
+        (true, None, (_: Option[String]) => Some(futureItemsSyncCursor))
+    }
+    val projectInstance = (nextCursor: Option[String]) =>
+      toProjectedInstance(_, toProjectedCursor(nextCursor), selectedProps)
+    val dataFetcher = (cursors: Option[Map[String, String]]) =>
+      fetchData(isBackFill, cursors, `with`, select)
+    syncOut(cursors, dataFetcher, projectInstance, applyTimestampTermination = !isBackFill)
+  }
+
+  private def syncOut(
+      cursors: Option[Map[String, String]],
+      fetchData: Option[Map[String, String]] => IO[ItemsWithCursor[InstanceDefinition]],
+      projectInstance: Option[String] => InstanceDefinition => ProjectedFlexibleDataModelInstance,
+      applyTimestampTermination: Boolean
+  ): Stream[IO, ProjectedFlexibleDataModelInstance] =
+    Stream.eval {
+      fetchData(cursors).map { sr =>
+        val items = sr.items
+        val nextCursor = sr.nextCursor
+        val shouldStopEarly = items.isEmpty || (applyTimestampTermination && items.last.lastUpdatedTime >=
```

Comment:
I think we may still have a problem here if /query may return empty page but non-empty next cursor and mean that we should page more (though somewhat likely that fdm api never returns an empty page with nextCursor present)

Reply:
ah.. for backfill we must not stop early I think.. 
We accept seeing duplicates on the next session..

Reply:
yeah, so now I've added a flag to disable early termination, but it doesn't disable termination on seeing an empty page, not sure what's safer here

Reply:
another version could be something like
```kotlin
val shouldStopByTimestamp = !isBacFill && items.last.lastUpdatedTime >= abc;
val shouldStopByEmptyPage = !isBackFill && items.isEmpty
val shouldStop = nextCursor.isEmpty || shouldStopByTimestamp || shouldStopByEmptyPage
```